src/Microsoft.ServiceFabric.Actors.Migration/AmbiguousActorIdHandler.cs

```diff
@@ -0,0 +1,260 @@
+            if (stringActorId.Contains("_"))
+            {
+                var lastIndexOfUnderscore = stringActorId.LastIndexOf('_');
+                tempStr = stringActorId.Substring(0, lastIndexOfUnderscore);
+            }
+
+            for (int iterations = 0; iterations < stringActorId.Split('_').Count(); iterations++)
+            {
+                count += this.metadataCache.Select(n => n.Contains(tempStr)).Count();
+
+                var lastIndexOfUnderscore = tempStr.LastIndexOf('_');
+                tempStr = tempStr.Substring(0, lastIndexOfUnderscore);
+            }
+
+            if (count > 1)
+            {
+                foreach (Type resolver in this.resolvers)
+                {
+                    var resolverObj = Activator.CreateInstance(resolver, new object[] { });
+                    var tryResolveActorIdAndStateNameMethodInfo = resolver.GetMethod("TryResolveActorIdAndStateName");
+                    var tryResolveActorIdAndReminderNameMethodInfo = resolver.GetMethod("TryResolveActorIdAndReminderName");
+
+                    string actorId = string.Empty;
+                    string stateName = string.Empty;
+                    string reminderName = string.Empty;
+
+                    if ((bool)tryResolveActorIdAndStateNameMethodInfo.Invoke(resolverObj, new object[] { stringActorId, actorId, stateName })
+                        || (bool)tryResolveActorIdAndReminderNameMethodInfo.Invoke(resolverObj, new object[] { stringActorId, actorId, reminderName }))
+                    {
+                        // actor Id is not ambigious as resolver was successful
+                        return false;
+                    }
+                }
+            }
+
+            return true;
+        }
+
+        internal ActorId GetActorId(string key)
+        {
+            if (!this.IsAmbiguousActorId(key))
+            {
+                var count = 0;
+                var tempStr = key;
+                var actorId = key;
+
+                // drop last token. which could be statename or reminder name
+                if (key.Contains("_"))
+                {
+                    var lastIndexOfUnderscore = key.LastIndexOf('_');
+                    tempStr = key.Substring(0, lastIndexOfUnderscore);
+                    actorId = tempStr;
+                }
+
+                for (int iterations = 0; iterations < actorId.Split('_').Count(); iterations++)
+                {
+                    count += this.metadataCache.Select(n => n.Contains(tempStr)).Count();
+
+                    var lastIndexOfUnderscore = tempStr.LastIndexOf('_');
+                    tempStr = tempStr.Substring(0, lastIndexOfUnderscore);
+                }
+
+                if (count <= 1)
+                {
+                    return new ActorId(actorId);
+                }
+                else
+                {
+                    foreach (Type resolver in this.resolvers)
```

Comment:
The cache is guaranteed to have non ambiguous actor ids, meaning only one actor id will match.